django/thunderstore/repository/models/package_version.py

```diff
@@ -310,6 +326,148 @@ def log_download_event(version_id: int, client_ip: Optional[str]):
+                    message=message,
+                )
+            )
+        else:
+            raise PermissionError()
+
+    def can_user_manage_approval_status(self, user: Optional[UserType]) -> bool:
+        if self.review_status == PackageVersionReviewStatus.immune:
+            return False
+
+        for listing in self.package.community_listings.all():
+            if listing.can_user_manage_approval_status(user):
+                return True
+        return False
+
+    def is_visible_to_user(self, user: UserType) -> bool:
+        if not self.visibility:
+            return False
+
+        if self.visibility.public_detail:
+            return True
+
+        if user is None:
+            return False
+
+        if self.visibility.owner_detail:
+            if self.package.owner.can_user_access(user):
+                return True
+
+        if self.visibility.moderator_detail:
+            for listing in self.package.community_listings.all():
+                if listing.community.can_user_manage_packages(user):
+                    return True
+
+        if self.visibility.admin_detail:
+            if user.is_superuser:
+                return True
+
+        return False
+
+    @transaction.atomic
+    def update_visibility(self):
+        """
+        Updates the version's visibility based on whether it or its package is active and its review status.
+        This may also change the visibility of related listings, so it calls update_visibility() on them.
+
+        By default, versions are visible to everyone (for now). Rejected versions aren't publicly visible,
+        and inactive versions or versions with inactive packages aren't visible at all.
+        """
+        original_visibility_bitstring = self.visibility.bitstring()
+
+        self.visibility.public_detail = True
+        self.visibility.public_list = True
+        self.visibility.owner_detail = True
+        self.visibility.owner_list = True
+        self.visibility.moderator_detail = True
+        self.visibility.moderator_list = True
+
+        if not self.is_active or not self.package.is_active:
+            self.visibility.public_detail = False
+            self.visibility.public_list = False
+            self.visibility.owner_detail = False
+            self.visibility.owner_list = False
+            self.visibility.moderator_detail = False
+            self.visibility.moderator_list = False
+
+        if (
+            self.review_status == PackageVersionReviewStatus.rejected
+            or self.review_status == PackageVersionReviewStatus.pending
+        ):
+            self.visibility.public_detail = False
+            self.visibility.public_list = False
+
+        if self.visibility.bitstring != original_visibility_bitstring:
+            self.visibility.save()
+            for listing in self.package.community_listings.all():
+                listing.update_visibility()
+            self.package.recache_latest()
```

Comment:
I feel like we should think about how to build this in a way that's easy to test and maintain, if nothing else by splitting it to more atomic components that are easier to write standalone tests for rather than having to test everything at once.

Right now it seems like we're actually dealing with a pretty simple map of `conditions that need to match` to `flags that should be flipped to False`. If that's a paradigm we can assume is true, this shouldn't be too hard to turn into a system of sorts that functions in a more "data driven" fashion, making it easier to reliably test.

Reply:
Should note that this approach does have some advantages (mainly simplicity to read), but I suspect a light abstraction layer could still prove useful long-term and can actually make it easier to reason about if you have the required contextual info. Reading through dozens of lines of code isn't as straight forward as seeing ("if the package is not active, visibility rules of inactive packages are applied") in a single line after all. What those rules are isn't as relevant on this level of context in my opinion